src/packet_analysis/protocol/ieee802_11/IEEE802_11.cc

```diff
@@ -106,5 +106,17 @@ bool IEEE802_11Analyzer::AnalyzePacket(size_t len, const uint8_t* data, Packet*
 	uint32_t protocol = (data[0] << 8) + data[1];
 	data += 2;
 
-	return ForwardPacket(len - len_80211, data, packet, protocol);
+	if ( packet->tunnel_type == BifEnum::Tunnel::NONE )
```

Comment:
I can't say I've grasped this. It's somehow relying on outer analyzer setting tunneling so we trigger some special logic below? That probably makes sense for the GRE/ARUBA setup?

Reply:
This is actually the impetus for that change in the default `tunnel_type`. We check the tunnel type here because we need to reset the protocol being passed down to `packet_analysis::IP::ParsePacket` (called by the `IPTunnel` analyzer). Otherwise it reports a bad protocol identifier.

In reverse, if the default `tunnel_type` is set to `IP` and not `NONE`, it will erroneously fall into the second if block here and try to call the tunnel analyzer for something that isn't a tunnel.

Reply:
TLDR: This special case here didn't click at all. It does a bit now, but I think it shouldn't be there, but that probably requires surgery.

This is sending the IP packet within 802.11 through the IPTunnel analyzer because there's an outer tunnel. That seems fairly surprising: Even if there's an outer tunnel, an IP packet within 802.11 should be sent to just the IP analyzer. The outer tunnel should not be relevant (and forwarding based on just the LLC protocol be "the right thing" outside of crazy things (and this doesn't seem too crazy).

So.. staring for a long time: I'm seeing that with this change, a tunnel.log entry for the outer GRE connection is created and that seems nice. If I patch to just dispatch on llc.protocol, that tunnel.log entry is gone and that's sad, I agree. Though that's pretty confusing and actually seems an issue on the level of the GRE analyzer implemenation: Conceptually, any IP/GRE/something tunnel should likely produce a tunnel.log entry as long as the IP and GRE parts are valid, even if the "something part" isn't something that Zeek can deal with.

I realize the current code is currently structured much differently and by default any inner protocols within GRE are sent through IPTUNNEL by default and that then happens to create the tunnel.log. This also made me understand where the special case here really originates from, but at the same time appears pretty rough.

I think the cleaner solution would be to push tunnel instantiation and resulting tunnel.log generation into GRE itself rather than doing it indirectly through the IPTUNNEL analyzer.

Reply:
The question is whether that should happen as part of this PR.

The current IPTunnelAnalyzer was born out of a desire to not repeat a bunch of code around handling the encapsulation and whatnot in each of the tunnel analyzers. Maybe it could be reworked to become a base class for the other tunnel analyzers, but it also needs to handle regular IP-in-IP tunnels as well.

Reply:
FWIW, I agree that the current approach is hard to follow and I like the suggested solutions. Moving this into another PR makes sense to me.

Reply:
> The question is whether that should happen as part of this PR.

No. I understand this was meant as a quick fix, but you asked for review :-)

I have one more idea/patch below that removes the IEEE802.11 special case. That makes it a bit more graspable for me and possibly clearer how to re-structure in the future. If you don't like it, I'll sign-off on whatever otherwise, you know this stuff much better :-)

> The current IPTunnelAnalyzer was born out of a desire to not repeat a bunch of code around handling the encapsulation and whatnot in each of the tunnel analyzers. Maybe it could be reworked to become a base class for the other tunnel analyzers, but it also needs to handle regular IP-in-IP tunnels as well.

Yep, after staring even longer a few things came together. The IPTUNNEL analyzer seems rather GRE specific, so that was initially very confusing (and it still is).

I've tried one more thing: Doing the register_packet_analyzer() of the ARUBA types for IPTUNNEL instead of GRE. That works out well and allows to remove the IEEE802.11 tunnel special case there (which IMO really stood out as bad). Now the setup is GRE->IPTUNNEL->IEEE802.11 and it seems to work for the new test cases added :crossed_fingers: 

This is the change.
https://github.com/zeek/zeek/commit/78bc32c97c0e9525b63ec460dceecad3d3c8d322

The one quirky part that is left is that we need to keep setting gre_link_type properly in GRE, otherwise IPTunnel uses DLT_RAW for the 802.11 packet within and that's not valid.

Maybe there's actually room for an extension to `register_packet_analyzer()` to pass the link_type during registration to enable that use-case and doesn't rely on whatever the analyzer before ended up setting in gre_link_type? Definitely future though.

```
PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IPTUNNEL, 0x8200, PacketAnalyzer::ANALYZER_IEEE802_11, PacketAnalyzer::LINK_TYPE_IEEE802_11);
```

Reply:
> What confuses me here is that the GRE types for Aruba will still end up in the configuration of the IPTunnel analyzer, right?

Yeah, this is the part that I'm not really happy with either. That said, not having the weird work around in the 802.11 analyzer is nicer. I committed Arne's version since it removes the work around, but this should probably all be reinvestigated.

Reply:
> > What confuses me here is that the GRE types for Aruba will still end up in the configuration of the IPTunnel analyzer, right?
>
> Yeah, this is the part that I'm not really happy with either. [...] but this should probably all be reinvestigated.

(Re-)moving the IP analyzer pieces into GRE directly so it becomes closer to VXLAN/Geneve (and closer to what Jan posted) would be reasonable. I'm not sure Aruba or e.g., Tranparent Ethernet Bridging (0x6558) need to be analyzer: there's nothing to be parsed - they'd be coded mappers from GRE to IEEE802.11 or Ethernet analyzer. Didn't seem there's anything fancy about them other than not fitting the GRE / iptunnel analyzer setup and that makes it all complicated.

I've approved, @timwoj , assuming you can figure out test failures etc :-)